### PR TITLE
Use an empty slice as default value for DNS, DNSSearch and DNSOptions.

### LINKED ISSUE #17680: "docker inspect -f dns" incompatibility between 1.8 and 1.9
As noted in #17026,

    docker inspect -f {{.HostConfig.Dns}} somecontainer

works in 1.8 (and earlier) and doesn't work in 1.9, while

    docker inspect -f {{.HostConfig.DNS}} somecontainer

works in 1.9 and doesn't work in 1.8.  

This is a pretty bad compatibility breakage for anyone with a script that fetches this setting.
Also I cannot see the change mentioned in the release notes.

### PR DESCRIPTION
**Pull request by calavera**
Fixes #17680 

It also makes sure that already created containers return an empty array on inspect responses.

I think this is the least bad option. I'm working on an issue to explain the current issues with the api and how we're going to solve them for 1.10

### CODE REVIEW

**Comment by tiborvass** [comment]
@calavera can you explain why this works?

**Comment by calavera** [comment]
**Short story**

life is complicated

**Long story**

1. We deserialize the inspect response on typed structs.
2. Before 1.8, the container struct included `Dns`, we changed its name to `DNS` to please go-lint, keeping the api field name as `Dns` with a json tag.
3. Since the go template is executed on the typed struct, ContainerJSON in this case, `{{ .HostConfig.Dns }}` became an unknown field and the templates fails to execute.

*Enter raw inspect:*

When a template execution fails on the typed struct, we fall back to a raw check. We do this because swarm inject some fields that are not in any of our typed structs. How does this work?

1. We deserialize the inspect response on an interface, effectively a `map[string]interface{}`.
2. We execute the template on this interface. In this case, the template succeeds, but it returns `<no value>` because the template cannot find the right value for a `null` field. Pre 1.9, the null value for Dns was `[]`, because you know the type when you execute the template on a typed struct, `[]string` in this case.
3. Go 1.4 also returns `<no value>` for fields that are not in the struct. I introduced a hack in https://github.com/docker/docker/pull/17284/files#diff-ffaac1d66bac3b1d6d8af9a52b0f118fR168 to fail with the original template error when the value of a field in the raw inspect is `<no value>`.

*The solution:*

What do we want?

the value of the DNS array when we use `{{.HostConfig.Dns}}`, being and `[]` when there is nothing configured.

The simplest solution:

We let the client follow the workflow described in the previous points, but we make sure the DNS value is never null in the api response by initializing it to an empty array.

**Comment by tiborvass** [comment]
Are the `DNS...` fields the only ones affected?

File: runconfig/parse.go [hunk 1]
```diff
@@ -348,9 +348,9 @@ func Parse(cmd *flag.FlagSet, args []string) (*Config, *HostConfig, *flag.FlagSe
 		PortBindings:      portBindings,
 		Links:             flLinks.GetAll(),
 		PublishAllPorts:   *flPublishAll,
-		DNS:               flDNS.GetAll(),
-		DNSSearch:         flDNSSearch.GetAll(),
-		DNSOptions:        flDNSOptions.GetAll(),
+		DNS:               flDNS.GetAllOrEmpty(),
```
Comment (tiborvass) [inline runconfig/parse.go]: Can we add a small comment here explaining why we need GetAllOrEmpty?

**Comment by calavera** [comment]
> Are the DNS... fields the only ones affected?

yes

File: runconfig/parse.go (hunk 1 above)
Comment (calavera) [inline runconfig/parse.go]: Added

**Comment by tiborvass** [comment]
@calavera perfect! LGTM

File: daemon/inspect.go [hunk 2]
```diff
@@ -134,6 +134,24 @@ func (daemon *Daemon) getInspectData(container *Container, size bool) (*types.Co
 		HostConfig:   &hostConfig,
 	}
 
+	// Make sure the dns fields are never nil.
+	// New containers don't ever have those fields nil,
+	// but pre created containers can still have those nil values.
+	// See https://github.com/docker/docker/pull/17779
+	// for a more detailed explanation on why we don't want that.
+	var emptySlice []string
```
Comment (aaronlehmann) [inline daemon/inspect.go]: Won't this slice be `nil` unless you allocate it with `make` or by setting it to a slice literal?

Comment (tiborvass) [inline daemon/inspect.go]: true, this is nil, you'd need `emptySlice := []string{}`

**Comment by cpuguy83** [comment]
haha, daemon/inspect.go:142:2: can probably use "var emptySlice []string" instead

File: daemon/inspect.go [hunk 3]
```diff
@@ -134,6 +134,24 @@ func (daemon *Daemon) getInspectData(container *Container, size bool) (*types.Co
 		HostConfig:   &hostConfig,
 	}
 
+	// Make sure the dns fields are never nil.
+	// New containers don't ever have those fields nil,
+	// but pre created containers can still have those nil values.
+	// See https://github.com/docker/docker/pull/17779
+	// for a more detailed explanation on why we don't want that.
+	emptySlice := make([]string, 0)
```
Comment (tiborvass) [inline daemon/inspect.go]: try `emptySlice := []string{}` would `make validate` still be angry?

**Comment by calavera** [comment]
I've changed the patch to initialize those fields correctly after loading the configuration from disk. I think it makes more sense to be always consistent with their value than initialize them only in the special case of inspect.

Please, take a look.

**Comment by tiborvass** [comment]
LGTM again

**Comment by runcom** [comment]
LGTM

**Pull request merged**